dex/elrond_dex_pair/src/lib.rs

```diff
@@ -433,32 +477,59 @@ pub trait Pair:
 
         reserve_token_in += &amount_in_optimal_after_fee;
         reserve_token_out -= &amount_out;
-
         self.update_reserves(&reserve_token_in, &reserve_token_out, &token_in, &token_out);
-        self.send_tokens(&token_out, &amount_out, &caller);
-        self.send_tokens(&token_in, &residuum, &caller);
+
+        // A swap should not decrease the value of K. Should either be greater or equal.
+        let new_k = self.calculate_k_for_reserves();
+        self.validate_k_invariant(&old_k, &new_k)?;
 
         //The transaction was made. We are left with $(fee) of $(token_in) as fee.
         if self.is_fee_enabled() {
-            self.send_fee(token_in, fee_amount);
+            self.send_fee(&token_in, fee_amount);
         }
 
-        // A swap should not decrease the value of K. Should either be greater or equal.
-        let new_k = self.calculate_k_for_reserves();
-        self.validate_k_invariant(&old_k, &new_k)?;
+        self.send_tokens(&token_out, &amount_out, &caller, &opt_accept_funds_func)?;
+        self.send_tokens(&token_in, &residuum, &caller, &opt_accept_funds_func)?;
 
         Ok(())
     }
 
-    #[inline]
-    fn send_tokens(&self, token: &TokenIdentifier, amount: &Self::BigUint, destination: &Address) {
+    fn send_tokens(
+        &self,
+        token: &TokenIdentifier,
+        amount: &Self::BigUint,
+        destination: &Address,
+        opt_accept_funds_func: &OptionalArg<BoxedBytes>,
+    ) -> SCResult<()> {
         if amount > &0 {
-            let _ = self.send().direct_esdt_via_transf_exec(
+            let (function, gas_limit) = match opt_accept_funds_func {
+                OptionalArg::Some(accept_funds_func) => (
+                    accept_funds_func.as_slice(),
+                    self.transfer_exec_gas_limit().get(),
+                ),
+                OptionalArg::None => {
+                    let no_func: &[u8] = &[];
+                    (no_func, 0u64)
+                }
+            };
+
+            let result = self.send().direct_esdt_execute(
```

Comment:
direct_esdt_execute is not good - this works intra shard only. You should use transfer and execute and should not care about the results - as those may be different from intra and cross shard. In case of cross shard there are no results.

Reply:
I understood that behind this function we actually have the transferAndExecute API so this should be ok. However you should not care about the results of one send.

Reply:
If the transaction fails on the sender part - the execution will stop anyway and completely revert to previous state.

Reply:
Fixed